File: algorithms/code/leetcode/lc002_add_two_numbers/lc002_add_two_numbers.py
# https://leetcode.com/problems/add-two-numbers/

import unittest
from typing import Optional


class ListNode:
    def __init__(
        self, val: int = 0, next_item: Optional['ListNode'] = None
    ) -> None:
        self.val = val
        self.next = next_item


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        handle = ListNode(0)
        handle.next = None
        tail = handle
        carry = 0
        n1: ListNode | None = l1
        n2: ListNode | None = l2
        while n1 is not None or n2 is not None or carry != 0:
            if n1 is None:
                value1 = 0
            else:
                value1 = n1.val
                n1 = n1.next
            if n2 is None:
                value2 = 0
            else:
                value2 = n2.val
                n2 = n2.next
            summed = carry + value1 + value2
            tail.next = ListNode(summed % 10)
            tail.next.next = None
            tail = tail.next
            carry = summed // 10
        assert handle.next is not None
        return handle.next


class TestCode(unittest.TestCase):
    @staticmethod
    def linked_to_listed(linked: ListNode | None) -> list[int]:
        listed: list[int] = []
        while linked is not None:
            listed.append(linked.val)
            linked = linked.next
        return listed

    @staticmethod
    def listed_to_linked(listed: list[int]) -> ListNode:
        linked: ListNode | None = None
        for value in listed[::-1]:
            linked = ListNode(value, linked)
        assert linked is not None
        return linked

    def test_example(self) -> None:
        array1 = [2, 4, 3]
        array2 = [5, 6, 4]
        linked = Solution().addTwoNumbers(
            TestCode.listed_to_linked(array1),
            TestCode.listed_to_linked(array2),
        )
        expected = [7, 0, 8]
        assert TestCode.linked_to_listed(linked) == expected

    def test_uneven(self) -> None:
        array1 = [2, 4]
        array2 = [5, 6, 4]
        linked = Solution().addTwoNumbers(
            TestCode.listed_to_linked(array1),
            TestCode.listed_to_linked(array2),
        )
        expected = [7, 0, 5]
        assert TestCode.linked_to_listed(linked) == expected

    def test_carry(self) -> None:
        array1 = [2, 4]
        array2 = [5, 6]
        linked = Solution().addTwoNumbers(
            TestCode.listed_to_linked(array1),
            TestCode.listed_to_linked(array2),
        )
        expected = [7, 0, 1]
        assert TestCode.linked_to_listed(linked) == expected

File: algorithms/code/leetcode/lc002_add_two_numbers/lc002_add_two_numbers.rs
// https://leetcode.com/problems/add-two-numbers/
// #linking

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

pub struct Solution;

impl Solution {
    pub fn add_two_numbers(
        mut l1: Option<Box<ListNode>>,
        mut l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut handle = Some(Box::new(ListNode::new(0)));
        let mut tail = &mut handle;
        let mut carry = 0;
        while l1.is_some() || l2.is_some() || carry != 0 {
            let mut value1 = 0;
            if let Some(node) = l1 {
                value1 = node.val;
                l1 = node.next;
            }
            let mut value2 = 0;
            if let Some(node) = l2 {
                value2 = node.val;
                l2 = node.next;
            }
            let summed = carry + value1 + value2;
            if let Some(ref mut node) = tail {
                node.next = Some(Box::new(ListNode {
                    val: summed % 10,
                    next: None,
                }));
                tail = &mut node.next;
            }
            carry = summed / 10;
        }
        handle.unwrap().next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn vector_to_list(vector: Vec<i32>) -> Option<Box<ListNode>> {
        let mut list: Option<Box<ListNode>> = None;
        for &value in vector.iter().rev() {
            let node = Box::new(ListNode {
                val: value,
                next: list,
            });
            list = Some(node);
        }
        list
    }

    fn list_to_vector(mut list: Option<Box<ListNode>>) -> Vec<i32> {
        let mut vector = Vec::new();
        while let Some(node) = list {
            vector.push(node.val);
            list = node.next;
        }
        vector
    }

    #[test]
    fn test_example() {
        let array1 = vec![2, 4, 3];
        let array2 = vec![5, 6, 4];
        let listed = Solution::add_two_numbers(vector_to_list(array1), vector_to_list(array2));
        let expected = vec![7, 0, 8];
        assert_eq!(list_to_vector(listed), expected);
    }

    #[test]
    fn test_uneven() {
        let array1 = vec![2, 4];
        let array2 = vec![5, 6, 4];
        let listed = Solution::add_two_numbers(vector_to_list(array1), vector_to_list(array2));
        let expected = vec![7, 0, 5];
        assert_eq!(list_to_vector(listed), expected);
    }

    #[test]
    fn test_carry() {
        let array1 = vec![2, 4];
        let array2 = vec![5, 6];
        let listed = Solution::add_two_numbers(vector_to_list(array1), vector_to_list(array2));
        let expected = vec![7, 0, 1];
        assert_eq!(list_to_vector(listed), expected);
    }
}
